kai/llm_io_handler.py

```diff
@@ -18,6 +19,43 @@
 LLM_RETRY_DELAY = 10
 
 
+def get_prompt(
+    model_provider: ModelProvider,
+    pb_vars: dict,
+    path_templates: str = PATH_TEMPLATES,
+    jinja_kwargs: dict = None,
+):
+    """
+    Generate a prompt using Jinja templates based on the provided model
+    provider, variable dictionary, and optional path templates and Jinja
+    arguments.
+    """
+
+    if jinja_kwargs is None:
+        jinja_kwargs = {}
+
+    jinja_kwargs = {
+        "loader": FileSystemLoader(path_templates),
+        "undefined": StrictUndefined,
+        "trim_blocks": True,
+        "lstrip_blocks": True,
+        "autoescape": True,
+        **jinja_kwargs,
+    }
+
+    jinja_env = Environment(**jinja_kwargs)  # trunk-ignore(bandit/B701)
+
+    if model_provider.template is not None:
+        return jinja_env.get_template(model_provider.template).render(pb_vars)
+
+    # Try to render the template corresponding to the model_id, fallback to
+    # main.jinja
+    try:
+        return jinja_env.get_template(model_provider.model_id).render(pb_vars)
+    except TemplateNotFound:
```

Comment:
Maybe put a debug log in here, just in case someone expects a specific model template to be selected so that they can see it's not being found. Other than that looks good :+1: